src/cloudai/workloads/ai_dynamo/ai_dynamo.sh

```diff
@@ -79,6 +85,74 @@ _parse_cli_pairs() {
   done
 }
 
+_set_backend_defaults() {
+  case "${dynamo_args["backend"]}" in
+    vllm)
+      :
+      ;;
+    sglang)
+      dynamo_args["prefill-cmd"]="python3 -m dynamo.sglang.worker"
+      dynamo_args["decode-cmd"]="python3 -m dynamo.sglang.decode_worker"
+      dynamo_args["ingress-cmd"]="python3 -m dynamo.frontend"
+      ;;
+    *)
+      log "ERROR: Unknown backend '${dynamo_args["backend"]}'"
+      exit 1
+      ;;
+  esac
+}
+
+_is_vllm() { [[ "${dynamo_args["backend"]}" == "vllm" ]]; }
+_is_sglang() { [[ "${dynamo_args["backend"]}" == "sglang" ]]; }
+
+_sync_num_nodes_from_section_args() {
+  if [[ -n "${prefill_args["--num-nodes"]:-}" ]]; then
+    dynamo_args["num-prefill-nodes"]="${prefill_args["--num-nodes"]}"
+  fi
+  if [[ -n "${decode_args["--num-nodes"]:-}" ]]; then
+    dynamo_args["num-decode-nodes"]="${decode_args["--num-nodes"]}"
+  fi
+}
+
+_csv_len() { grep -oE '[^,]+' <<< "$1" | wc -l; }
+
+_first_in_csv() { echo "$1" | cut -d',' -f1; }
+
+_csv_index_of() {
+  local list="$1" name="$2"
+  local IFS=',' arr i
+  read -ra arr <<< "$list"
+  for i in "${!arr[@]}"; do
+    if [[ "${arr[$i]}" == "$name" || "${arr[$i]}" == *"$name"* || "$name" == *"${arr[$i]}"* ]]; then
+      echo "$i"; return 0
+    fi
+  done
+  echo "-1"
+}
+
+_validate_or_build_nodelists() {
+  local dl_len=$(_csv_len "${dynamo_args["decode-nodelist"]}")
+  local pl_len=$(_csv_len "${dynamo_args["prefill-nodelist"]}")
+  if (( dl_len > 0 )); then dynamo_args["num-decode-nodes"]="$dl_len"; fi
+  if (( pl_len > 0 )); then dynamo_args["num-prefill-nodes"]="$pl_len"; fi
+
+  if [[ -z "${dynamo_args["decode-nodelist"]}" || -z "${dynamo_args["prefill-nodelist"]}" ]]; then
+    if [[ -z "${DYNAMO_NODELIST:-}" ]]; then
+      log "ERROR: Provide --dynamo-decode-nodelist/--dynamo-prefill-nodelist or set DYNAMO_NODELIST"; exit 1
+    fi
+    local d="${dynamo_args["num-decode-nodes"]}"
+    local p="${dynamo_args["num-prefill-nodes"]}"
+    local total=$(_csv_len "${DYNAMO_NODELIST}")
+    if (( total < d + p )); then
+      log "ERROR: DYNAMO_NODELIST has ${total} entries; need decode(${d})+prefill(${p})"; exit 1
+    fi
+    [[ -z "${dynamo_args["decode-nodelist"]}" ]] && \
+      dynamo_args["decode-nodelist"]=$(echo "$DYNAMO_NODELIST" | cut -d',' -f1-"$d")
+    [[ -z "${dynamo_args["prefill-nodelist"]}" ]] && \
+      dynamo_args["prefill-nodelist"]=$(echo "$DYNAMO_NODELIST" | cut -d',' -f$(( d + 1 ))-)
```

Comment:
The above stricter requirement will make this logic cleaner. Otherwise we might run into weird corner cases.